src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/CreateOrUpdateKubeBase.cs

```diff
@@ -0,0 +1,363 @@
+            });
+
+            if (!success)
+            {
+                throw new CmdletInvocationException(
+                    "Could not create a service principal with the right permissions. Are you an Owner on this project?");
+            }
+
+            AddRoleAssignment("Contributor", app.AppId);
+        }
+
+        protected bool Exists()
+        {
+            try
+            {
+                var exists = Client.ManagedClusters.Get(ResourceGroupName, Name) != null;
+                WriteVerbose(string.Format("Cluster exists: {0}", exists));
+                return exists;
+            }
+            catch (CloudException)
+            {
+                WriteVerbose("Cluster does not exist.");
+                return false;
+            }
+        }
+
+        protected void AddRoleAssignment(string role, string appId)
+        {
+            var success = RetryAction(() =>
+                AuthClient.RoleAssignments.Create(role, appId, new RoleAssignmentCreateParameters()
+                {
+                    Properties = new RoleAssignmentProperties(role, appId)
+                }));
+
+            if (!success)
+            {
+                throw new CmdletInvocationException(
+                    "Could not create a service principal with the right permissions. Are you an Owner on this project?");
+            }
+        }
+
+        protected static bool RetryAction(Action action)
+        {
+            var success = false;
+            foreach (var i in Enumerable.Range(1, 10))
+            {
+                try
+                {
+                    action();
+                    success = true;
+                    break;
+                }
+                catch (Exception)
+                {
+                    // AAD might puke here, so we catch it and try again until success
+                    Thread.Sleep(1000 * i);
+                }
+            }
+            return success;
+        }
+
+        protected AcsServicePrincipal LoadServicePrincipal()
+        {
+            var config = LoadServicePrincipals();
+            return config?[DefaultProfile.DefaultContext.Subscription.Id];
+        }
+
+        protected Dictionary<string, AcsServicePrincipal> LoadServicePrincipals()
+        {
+            return File.Exists(AcsSpFilePath)
+                ? JsonConvert.DeserializeObject<Dictionary<string, AcsServicePrincipal>>(
+                    File.ReadAllText(AcsSpFilePath))
+                : null;
+        }
+
+        protected void StoreServicePrincipal(AcsServicePrincipal acsServicePrincipal)
+        {
+            var config = LoadServicePrincipals() ?? new Dictionary<string, AcsServicePrincipal>();
+            config[DefaultProfile.DefaultContext.Subscription.Id] = acsServicePrincipal;
+            File.WriteAllText(AcsSpFilePath, JsonConvert.SerializeObject(config));
+        }
+
+        protected static string RandomUtfString(int size)
```

Comment:
Many of these seems like utility functions that should be factored into a utility class, or implementd as extensions over the base cmdlet type.